allocator.go

```diff
@@ -82,27 +83,39 @@ func (a *Allocator) literal(r Range, n Node) cpsatsolver.Literal {
 	return a.assignment[r.id][n.id]
 }
 
-func (a *Allocator) Allocate() (ok bool, assignments map[RangeID][]NodeID) {
-	for _, r := range a.ranges {
-		a.model.AddConstraints(cpsatsolver.NewExactlyKConstraint(r.rf, a.rangeLiterals(r)...))
+func WithNodeCapacityConstraint() ConstraintOption {
+	return func(al *Allocator) {
+		al.adhereToNodeResourcesConstraint()
```

Comment:
Instead of invoking the method right away, add an options type on Allocator.
```
type Allocator struct {
   // ...
   options struct {
      WithNodeCapacity bool
   }
}
```

In `Allocate`, you can later check:
```
func (a *Allocator) Allocate(...) {
    //
    if a.options.WithNodeCapacity {
         // ...
    }
}
```

```
func WithNodeCapacityConstraint() Option {
    return func (a *Allocator) {
        a.options.WithNodeCapacity = true
    }
}
```

This is a bit better because with variadic arguments it's possible to pass the same option twice, or to pass incompatible options. So you don't want to "execute" the options right away, you want to stage everything as part of this options struct and possible validate + return an informative error.

Reply:
Great call-out, I was debating with myself over the existing approach and the one you already outlined, the reason I chose the former is that I thought if we have `n` options, then that translates into `n` conditionals inside the `Allocate()` method. However, I agree that the passing of incompatible options is definitely an issue, repeated args not so much since duplicates should not change anything. I'll go with the approach you've suggested and then we can see how things pan out later on. Thanks!